File: idn-ideation-system/frontend/src/types/graph.ts
export type NodeType = 'intro' | 'scene' | 'choice' | 'ending' | 'question'
export type NarrativeFormat = 'branching' | 'hypertext'

export interface StoryChoice {
  id: string
  label: string
  targetNodeId: string | null
}

export interface StoryNodeData {
  title: string
  body: string
  description?: string     // short author description (not shown to audience)
  personalNote?: string    // private author notes
  nodeType: NodeType
  tags: string[]
  aiGenerated: boolean
  pending?: boolean
  navigationMode?: 'next' | 'choices'   // for scene nodes
  choices?: StoryChoice[]               // for choice nodes
}

export interface StoryNode {
  id: string
  type: 'storyNode'
  position: { x: number; y: number }
  data: StoryNodeData
}

export interface StoryEdge {
  id: string
  source: string
  target: string
  label: string
  type?: string
}

// Returned by the generate-narrative endpoint before conversion to StoryNode/StoryEdge
export interface GeneratedNode {
  id: string
  title: string
  body: string
  node_type: string
  tags: string[]
  x: number
  y: number
}

export interface GeneratedEdge {
  id: string
  source: string
  target: string
  label: string
}

File: idn-ideation-system/backend/models/graph.py
from enum import Enum
from typing import Any, Optional
from pydantic import BaseModel, ConfigDict


class NodeType(str, Enum):
    intro = "intro"
    scene = "scene"
    choice = "choice"
    ending = "ending"
    note = "note"  # kept for backward-compat with saved projects


class StoryNodeData(BaseModel):
    model_config = ConfigDict(extra="allow")

    title: str
    body: str = ""
    description: str = ""
    personal_note: str = ""
    node_type: NodeType = NodeType.scene
    tags: list[str] = []
    ai_generated: bool = False
    navigation_mode: Optional[str] = None
    choices: Optional[list[Any]] = None


class Position(BaseModel):
    x: float
    y: float


class StoryNode(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: str
    type: str = "storyNode"
    position: Position
    data: StoryNodeData


class StoryEdge(BaseModel):
    id: str
    source: str
    target: str
    label: str = ""
    type: str = "default"
